scipy/optimize/_optimize.py

```diff
@@ -3477,12 +3489,23 @@ def brute(func, ranges, args=(), Ns=20, full_output=0, finish=fmin,
     for k in range(N):
         if type(lrange[k]) is not type(slice(None)):
             if len(lrange[k]) < 3:
-                lrange[k] = tuple(lrange[k]) + (complex(Ns),)
+                if integrality[k]:
```

Comment:
All of the following checks (and my suggested changes) are ignored if the user supplies a `slice` for a given parameter.
The grid that's evaluated should not possess integer values that:

1. lie outside the slice start/stop values (low, high values).
2. are repeated
3. if there are no integer values that lie between the slice start/stop then an Exception should be raised.

Reply:
Yeah, my thought was that if the user passes a slice, it completely defines the choice of values, so we shouldn't try to change it. And since there wasn't much input validation in this function to begin with, I didn't try to intervene because `slice` objects are not super fun to work with. 

Do you mean that we should modify the slice (e.g. to prevent repeating integer values) or just add validation?

(Side note: does it also seem unusual to you to use slices as an input? Maybe e.g. an _array_ of values should be allowed?)

Reply:
If integrality is applied for user-supplied slices then I do think we should enforce bounds, at least one integer, and uniqueness.

Personally I don't use `slice` all that much. I guess it's a cheap way to specify a spaced array: `range, np.linspace, np.arange, np.geomspace`. It would be nice to have the capability to be able to specify arrays as well, so you could have your own grid.

Reply:
I'm not treating slices and tuples the same way; both grids assume only integer values.